### DISCUSSION TITLE
CI/CD bot support for GitLab

### DISCUSSION THREAD

**Discussion by almereyda** [TobikoData/sqlmesh#2372] [Ideas]
It could be nice to build MR-based DataOps workflows in GitLab.

Is this a considerable option? I've seen that the CI/CD bot already knows about a `type: github`, which invites for additions. Would `type: gitlab` be generally favoured and considered useful to be included in the roadmap?

**Comment by isaiahgm** [comment]
Second this, GitLab support is blocking us from using SQL Mesh to its fullest potential. Our company exclusively hosts code on GitLab, so the GitHub CI/CD isn't an option.

**Comment by almereyda** [comment]
Thanks to the well structured, modular and documented code in a self-contained repository, we find all artifacts to adapt in cleanly separated places:

- [sqlmesh/docs/integrations/github at main · TobikoData/sqlmesh](https://github.com/TobikoData/sqlmesh/tree/main/docs/integrations/github)
- [sqlmesh/sqlmesh/cicd at main · TobikoData/sqlmesh](https://github.com/TobikoData/sqlmesh/tree/main/sqlmesh/cicd)
- [sqlmesh/sqlmesh/integrations/github/cicd at main · TobikoData/sqlmesh](https://github.com/TobikoData/sqlmesh/tree/main/sqlmesh/integrations/github/cicd)
- [sqlmesh/tests/integrations/github/cicd at main · TobikoData/sqlmesh](https://github.com/TobikoData/sqlmesh/tree/main/tests/integrations/github/cicd)

How much effort would we assume it takes to adapt all this to the GitLab GraphQL API? Could the work be distributed across interested developers? While screenshots for the docs can only be generated after the fact, it may be useful to consider writing the tests first and adapting the CI/CD bot and the `gitlab` integration from the perspective of the expected API surface?

**Comment by eakmanrq** [comment]
>Is this a considerable option? I've seen that the CI/CD bot already knows about a type: github, which invites for additions.

Yes that is correct that this was designed to help facilitate more bots in the future. Definitely open to contributions to add bots for other platforms. It would require some extensive tests though (similar to the current bot) in order to help with the support burden of having these different bots.

>How much effort would we assume it takes to adapt all this to the GitLab GraphQL API?

Sadly I'm just not familiar with the Gitlab GraphQL API to have a sense on that complexity. Sadly I'm also not familiar with Gitlab at all since I've never used it. So this is the biggest issue for me personally to estimate the complexity involved. I'm also curious what features from the Github bot would also work on Gitlab given the constraints of that platform. 

>it may be useful to consider writing the tests first and adapting the CI/CD bot and the gitlab integration from the perspective of the expected API surface?

Yeah I think this is a good idea. I think the task might be finding each API request in the Github bot and then mapping that to a corresponding gitlab endpoint. That would also let us know if there is any functionality gap in Gitlab that we should be aware of before investing further. Then tests could be written to mock out those endpoints and verify the expected behavior. 

I think this would be great contribution to the SQLMesh project. The two key things that I would want to make sure we are tracking are:
* What are the functionality differences between the two bots and is the functionality that is possible on Gitlab enough to justify supporting it?
* How can we minimize the support burden for OSS developers and not require deep Gitlab expertise to do standard maintenance improvements to the bot over time?

**Reply by isaiahgm** [reply]
I'd be happy to contribute some time to help develop this, but I am also pretty unfamiliar with developing gitlab bots. My experience ends at writing simple CI/CD scripts for GitLab.

**Comment by riordan** [comment]
Has anyone had the chance to look into Gitlab support? Taking the path of Github & Airflow is absolutely the right path to adoption but as a Gitlab & Dagster user, I'd love to get the full SQLMesh experience too.

**Reply by tobymao** [reply]
We don’t currently have plans for gitlab, but dagster support will be a part of Tobiko Cloud. We do offer feature prioritization as a part of our Cloud / Enterprise offering. Please contact me on Slack if you’re interested in discussing.

**Comment by almereyda** [comment]
The mappings between the two by two API surfaces, depending on needs and support, seems doable.

- [GitHub REST API documentation - GitHub Docs](https://docs.github.com/en/rest?apiVersion=2022-11-28)
- [GitHub GraphQL API documentation - GitHub Docs](https://docs.github.com/en/graphql)
- [GraphQL API | GitLab](https://docs.gitlab.com/ee/api/graphql/)
- [REST API | GitLab](https://docs.gitlab.com/ee/api/rest/)

A quick check with the one deep link into the GitHub API docs within the bot controller to https://docs.github.com/en/graphql/reference/enums#mergestatestatus has revealed https://docs.gitlab.com/ee/api/graphql/reference/#detailedmergestatus. This gives confidence that also edge features might as well just be supported.

Glancing over the list of symbols in that file that GitHub offers

![grafik](https://github.com/user-attachments/assets/f725b4ae-39f2-40be-87a0-a30059b1101d)

`logger`, `TestFailure` and `BotCommand` appear generic enough not to depend on GitHub at all, leaving the rest for porting. A community-driven GitLab Python API client implementation exists, including experimental support for GraphQL:

- [python-gitlab/python-gitlab: A python wrapper for the GitLab API.](https://github.com/python-gitlab/python-gitlab)
  - [Using the REST API - python-gitlab v5.6.0](https://python-gitlab.readthedocs.io/en/stable/api-usage.html)
  - [Using the GraphQL API (beta) - python-gitlab v5.6.0](https://python-gitlab.readthedocs.io/en/stable/api-usage-graphql.html)

Looks well worth a try. Maybe we discover the API surfaces and mechanics overlap widely and a lot of code can be decoupled, potentially "only" swapping API clients here and there.

Leaving these thoughts here, in case I return here later again.
